### PR TITLE
Add Aceex adapter

### PR DESCRIPTION
**Pull request by And1sS**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final Imp firstImp = request.getImp().get(0);
+        try {
+            final ExtImpAceex extImpAceex = mapper.mapper().convertValue(
+                    firstImp.getExt(), ACEEX_EXT_TYPE_REFERENCE).getBidder();
+            return Result.withValue(makeHttpRequest(request, extImpAceex.getAccountId()));
+        } catch (DecodeException e) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:56]: IllegalArgument

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            return Result.withError(BidderError.badInput("Ext.bidder not provided"));
+        }
+    }
+
+    private HttpRequest<BidRequest> makeHttpRequest(BidRequest request, String accountId) {
+        return HttpRequest.<BidRequest>builder()
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:62]: Let's inline this method

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                .method(HttpMethod.POST)
+                .uri(resolveEndpoint(accountId))
+                .headers(constructHeaders(request))
+                .body(mapper.encode(request))
+                .payload(request)
+                .build();
+    }
+
+    private static MultiMap constructHeaders(BidRequest bidRequest) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:71]: Pls change order of `constructHeaders` and `resolveEndpoint`

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        final List<SeatBid> seatBids = ObjectUtils.getIfNotNull(bidResponse, BidResponse::getSeatbid);
+        final SeatBid firstSeatBid = CollectionUtils.isNotEmpty(seatBids) ? seatBids.get(0) : null;
+        if (firstSeatBid == null) {
+            throw new PreBidException("Empty SeatBid array");
+        }
+
+        final List<Bid> bids = ListUtils.emptyIfNull(firstSeatBid.getBid());
+        return bidsFromSeatBid(bids, bidRequest, bidResponse);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:110]: Inline 
```java
        return CollectionUtils.emptyIfNull(firstSeatBid.getBid()).stream()
                .filter(Objects::nonNull)
                .map(bid -> BidderBid.of(bid,
                        getBidMediaType(bid.getImpid(), bidRequest.getImp()), bidResponse.getCur()))
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
```

File: src/main/java/org/prebid/server/util/HttpUtil.java [hunk 5]
```diff
@@ -129,7 +129,7 @@ public static MultiMap headers() {
      */
     public static void addHeaderIfValueIsNotEmpty(MultiMap headers, CharSequence headerName, CharSequence headerValue) {
         if (StringUtils.isNotEmpty(headerValue)) {
-            headers.add(headerName, headerValue);
+            headers.set(headerName, headerValue);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/util/HttpUtil.java:132]: There are headers that can have multiple entries, if we will talk about XFF, it's also a header that can contain multiple values.
But if it's related bidder code, and u want only one value, pls do it in bidder

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final Imp firstImp = request.getImp().get(0);
+        try {
+            final ExtImpAceex extImpAceex = mapper.mapper().convertValue(
+                    firstImp.getExt(), ACEEX_EXT_TYPE_REFERENCE).getBidder();
+            return Result.withValue(makeHttpRequest(request, extImpAceex.getAccountId()));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badInput("Ext.bidder not provided"));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:57]: Pls, add test for this case

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/util/HttpUtil.java (hunk 5 above)
Comment (And1sS) [reply src/main/java/org/prebid/server/util/HttpUtil.java:132]: Ok, then maybe i will add **setHeaderIfValueIsNotEmpty** utility method?

File: src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        final SeatBid firstSeatBid = CollectionUtils.isNotEmpty(seatBids) ? seatBids.get(0) : null;
+        if (firstSeatBid == null) {
+            throw new PreBidException("Empty SeatBid array");
+        }
+
+        return CollectionUtils.emptyIfNull(firstSeatBid.getBid()).stream()
+                .filter(Objects::nonNull)
+                .map(bid -> BidderBid.of(bid, getBidMediaType(bid.getImpid(), bidRequest.getImp()),
+                        bidResponse.getCur()))
+                .filter(Objects::nonNull)
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/aceex/AceexBidder.java:109]: This one seems redundant

File: src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        // when
+        final Result<List<BidderBid>> result = aceexBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getValue()).isEmpty();
+        assertThat(result.getErrors()).hasSize(1)
+                .allSatisfy(error -> {
+                    assertThat(error.getType()).isEqualTo(BidderError.Type.bad_server_response);
+                    assertThat(error.getMessage()).startsWith("Empty SeatBid array");
+                });
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java:143]: Just containsExactly will be enough, this message doesn't need to be checked as startsWith

File: src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        // when
+        final Result<List<BidderBid>> result = aceexBidder.makeBids(httpCall, null);
+
+        // then
+        assertThat(result.getValue()).isEmpty();
+        assertThat(result.getErrors()).hasSize(1)
+                .allSatisfy(error -> {
+                    assertThat(error.getType()).isEqualTo(BidderError.Type.bad_server_response);
+                    assertThat(error.getMessage()).startsWith("Bad Server Response");
+                });
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java:125]: Contains exactly enough

File: src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+    public void makeHttpRequestsShouldReturnErrorsOfNotValidImps() {
+        // given
+        final BidRequest bidRequest = givenBidRequest(
+                impBuilder -> impBuilder
+                        .ext(mapper.valueToTree(ExtPrebid.of(null, mapper.createArrayNode()))));
+        // when
+        final Result<List<HttpRequest<BidRequest>>> result = aceexBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).hasSize(1);
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/aceex/AceexBidderTest.java:68]: Size check redundant

**Review by snahornyi — CHANGES_REQUESTED** [review]

**Review by snahornyi — APPROVED** [review]

**Pull request merged**